### DISCUSSION TITLE
[OAuth2] How to limit access by 'scope' while setting 'global_credentials' to true

### DISCUSSION THREAD

**Discussion by mong0520** [Kong/kong#9476] [Help]
Hi forks,

Suppose I have two services named A and B, both of them have OAuth2 plugin enabled and set `global_credentials` as true.

As I know, a token issued by service B can access service A (that’s what ‘global_credentials’ do)

But my question is that – suppose the defined scope of A is `user_profile`, and defined scopes of B are `email, user_profile`.
How should I reject the request to service A with token issued by B and the scope is only ‘email’ ?

**Comment by hanshuebner** [comment]
Hi @mong0520,

there is no inherent mechanism in the OAuth2 plugin that would allow matching the scopes of a global token on the service level.  The scopes are apparently verified only when creating the access token.  I think this is a bug that needs to be addressed in the OAuth2 plugin and can't really be worked around.  I have filed an internal ticket to get this addressed.

Thank you for the report!

Kind regards,
Hans

**Comment by vm-001** [comment]
Hi @mong0520 

The current OAuth2 plugin does support limiting access by 'scope'. You probably can have another layer(by writing a custom plugin), enable your custom plugin alone with those global oauth2 plugins. 

```
ServiceA
  plugins:
    - name: oauth2
       config:
         global_credentials: true,
         scopes: ["user_profile"]
    - name: your-custom-plugin
       config:
         require_scope: ["user_profile"]

ServiceB
  plugins:
    - name: oauth2
       config:
         global_credentials: true,
         scopes: ["email", "user_profile"]
    - name: your-custom-plugin:
       config:
         require_scope: ["email", "user_profile"]
```

In your-custom-plugin, retrieve the `X-Authenticated-Scope` header added by OAuth2, and then do your custom access limiting based on the plugin configuration.
